UI for setting the reason why SHF is waiting on an applicant

weedySeaDragon opened the pull request:
PT Story:  https://www.pivotaltracker.com/story/show/143810729

This is the 2nd PR (the 2nd half) for the story.

Changes proposed in this pull request:
1.  Seed the db with 1 reason: the "other/custom" reason because it is used specifically.  (Other reasons can be entered via the usual web pages/interface).
2. UI: the _reasons for waiting_ has been moved to above the _application status buttons
3.  Uses AJAX to update the db immediately when a reason is selected from the list.  The db is also updated immediately when the text for a 'custom/other' reason is entered.

### Screenshots (Optional):

<img width="690" alt="reasonwaiting-select-reason-sv" src="https://user-images.githubusercontent.com/673794/27395498-dc50d784-5664-11e7-839b-a05a26631f05.png">


<img width="843" alt="reasonwaiting-select-reason-en" src="https://user-images.githubusercontent.com/673794/27395496-dc495252-5664-11e7-8d58-1a100d4f5b09.png">

---

#### The reason isn't shown if the status is not 'Waiting for Applicant'
<img width="623" alt="reasonwaiting-hidden-unless-status-en" src="https://user-images.githubusercontent.com/673794/27395497-dc49d9a2-5664-11e7-8ce8-74a0a92002da.png">

---

#### The text field for typing in the "other reason" only shows if that reason is selected

<img width="845" alt="reasonwaiting-other-reason-text-en" src="https://user-images.githubusercontent.com/673794/27395495-dc4919cc-5664-11e7-9919-3b4443db5264.png">


Ready for review:
@patmbolger @thesuss @RobertCram

weedySeaDragon committed:
update ruby_version -> 2.4.0

weedySeaDragon committed:
Merge branch 'develop' into sprint16-#143810729-UI-custom-reason-waiting

# Conflicts:
#	.ruby-version

weedySeaDragon committed:
add attributes to MembershipApplication: waiting_reason & custom_reason_text; migrate db

weedySeaDragon committed:
seed MemberAppWaitingReason with the "Other/Custom" reason (text from I18n locales)

weedySeaDragon committed:
pass locale to other_reason_name/_desc to make seeding easier

weedySeaDragon committed:
features & specs for setting reasons for waiting

weedySeaDragon committed:
remove assertion no longer used (removed the 'submit/save' button)

weedySeaDragon committed:
helpers: return the correct method to call based on the current local

weedySeaDragon committed:
add allowed attributes to Pundit policy

weedySeaDragon committed:
refactor reasons_waiting into a parital; move to above the 'status' buttons/sections

weedySeaDragon committed:
respond to a .js request for the reason waiting

weedySeaDragon committed:
I18n locale entries needed

weedySeaDragon committed:
if Seeding tests are run prior to this spec, ensure the table is cleared out

weedySeaDragon committed:
CSS styling

weedySeaDragon committed:
Merge branch 'develop' into sprint16-#143810729-UI-custom-reason-waiting

# Conflicts:
#	db/seeds.rb

weedySeaDragon committed:
fix: seed the EN name of the 'other reason' (should have passed in :en locale)

weedySeaDragon committed:
need to get the current text for the 'other reason' if the locale has changed (not just when page loaded)

weedySeaDragon committed:
Merge branch 'develop' into sprint16-#143810729-UI-custom-reason-waiting

weedySeaDragon committed:
clean up per HoundCI suggestions

weedySeaDragon committed:
DRY error situations

RobertCram commented on app/models/admin_only/member_app_waiting_reason.rb:
```diff
@@ -12,15 +12,25 @@ def other_reason_placeholder?
-    def self.other_reason_name
-      I18n.t('admin_only.member_app_waiting_reasons.other_custom_reason')
+    def self.other_reason_name(locale = I18n.locale)
+      I18n.t('admin_only.member_app_waiting_reasons.other_custom_reason', locale)
```
The second parameter to the translate method should be an options hash. As it stands now the locale is ignored, and the method always returns the swedish translation

I18n.t('admin_only.member_app_waiting_reasons.other_custom_reason', locale: locale)

RobertCram commented on app/models/admin_only/member_app_waiting_reason.rb:
```diff
@@ -12,15 +12,25 @@ def other_reason_placeholder?
 
+    def self.other_reason_desc(locale = I18n.locale)
+      I18n.t('admin_only.member_app_waiting_reasons.other_custom_reason_desc', locale)
+    end
```
Same as above.

RobertCram commented:
Great work, Ashley. Apart from my inline comment, two minor remarks:

1. There should probably be some sort of protection against changing or deleting the "other/custom reason" record?

2. In the UI, I would personally prefer the "Other (enter the reason)" to be the last option in the list

weedySeaDragon committed:
features: other/custom reason comes from the locale file, is not a model (not in the db)

weedySeaDragon committed:
do not seed the 'custom/other' reason; it's not stored in the db

weedySeaDragon committed:
add set: select an item without any I18n

weedySeaDragon committed:
CSS styling

weedySeaDragon committed:
return "OK" if update successful; set the text for the "other/custom.." reason option

weedySeaDragon committed:
no .js is needed to return if action was successful

weedySeaDragon committed:
other_reason placeholder stuff no longer needed in Model; remove spec/tests

weedySeaDragon committed:
all logic for "enter custom reason.." into this javascript

weedySeaDragon committed:
Merge branch 'develop' into sprint16-#143810729-UI-custom-reason-waiting

weedySeaDragon commented:
I know the build fails.  I can't figure out why the cucumber feature tests are failing.   Any help is appreciated!

fyi:  the feature that fail is this file: `features/membership-application-status/admin-custom-reason-waiting.feature`

patmbolger committed:
interim progress (Pat)

(cherry picked from commit 75f5e10)

patmbolger committed:
refactored some JS

(cherry picked from commit 08ae9ba)

patmbolger committed:
changed feature steps in accordance with id changes in view (Pat)

(cherry picked from commit bc1d1be)

patmbolger committed:
removed selenium webdriver from Gemfile (Pat)

(cherry picked from commit aa7f58b)

patmbolger committed:
fixed a couple cuke steps (Pat)

(cherry picked from commit 2bf19ed)

patmbolger committed:
fixed remaining cuke tests (Pat)

(cherry picked from commit 560b1ca)

weedySeaDragon committed:
changes to CSS per houndci

weedySeaDragon commented:
passing locally on my machine with selenium, but not here with semaphore. <sigh>

weedySeaDragon committed:
POLTERGEIST fix: do not use "let" (use var instead)

https://github.com/teampoltergeist/poltergeist/issues/895

 Unfortunately parse time errors for ES6 features like *let* can be silently swallowed by PhantomJS, so you end up with no errors but also with no JS being actually run.

weedySeaDragon committed:
remove selenium (no longer needed)

weedySeaDragon committed:
update gemfile.lock

weedySeaDragon commented:
Closing this. Will replace it with with a much cleaner approach that @patmbolger came up with. 👍 

We sure did learn a lot about poltergeist, phantom, selenium, and `data-remote` (oh my!).  
 Hopefully we'll write up / document that.  Meanwhile, where's a short list:

### Lessons learned:
- phantom.js will silently fail if you use `let` in your Javascript (ECMS 6.0).  (Use 'var' instead)  Ref: [poltergeist issue 823: use of JavaScript's let will fail silently ](https://github.com/teampoltergeist/poltergeist/issues/823)
- waiting for a  "Then I should not see"  used `expect(page).not_to have_content content`.  However, this does _not_ make Capybara wait for ajax, which leads to all kind of messes (rails still running while the test is finished, etc.). (This made some of the tests fail due to timeout because they were waiting forever.)
  use Capybara's `assert_text` instead

- use `data-remote` to send info back to the controller, even if the parameters aren't what is 'normally' expected by the Controller.  Let the controller parse and deal with whatever parameters are sent.  Dealing with them in your Javascript is ugly and harder to maintain.

patmbolger commented:
Good list! 

In the 2nd item, the `assert_` Capybara methods do not wait on outstanding AJAX requests to complete, but they _do_ wait as long as the Capybara timeout setting (which I think is 10 seconds by default).  This is usually a much better approach than using Rspec `expect`.

A 4th item is the mechanism that @RobertCram developed to tell Capybara to wait until outstanding AJAX requests have all completed - that will prove to be useful, I am sure.  We should make that a separate test step altogether, so we can use it as needed (e.g. `Then I wait until AJAX requests complete`).  (kind of wonky, but since we don't expose the cute tests to the client I suppose we can get away with it).
